src/states/node.rs

```diff
@@ -2283,7 +2277,59 @@ impl Node {
             }
             outbox.send_event(Event::Terminate);
         }
-        false
+    }
+
+    // Drop peers to which we think we have a direct or tunnel connection, but where Crust reports
+    // that we're not connected to the peer or tunnel node respectively.
+    fn purge_invalid_rt_entries(&mut self, outbox: &mut EventBox) -> Transition {
+        let mut peer_ids_to_drop = vec![];
+        for (peer_id, name, is_tunnel) in self.peer_mgr.get_routing_peer_details() {
+            if is_tunnel {
+                match self.tunnels.tunnel_for(&peer_id) {
+                    Some(tunnel_node_id) => {
+                        if !self.crust_service.is_connected(tunnel_node_id) {
+                            debug!("{:?} Should have a tunnel connection to {} {:?} via {:?}, \
+                                    but tunnel node not connected.",
+                                   self,
+                                   name,
+                                   peer_id,
+                                   tunnel_node_id);
+                            peer_ids_to_drop.push(*tunnel_node_id);
+                        }
+                    }
+                    None => {
+                        if self.crust_service.is_connected(&peer_id) {
+                            debug!("{:?} Should have a tunnel connection to {} {:?}, but instead \
+                                    have a direct connection.",
+                                   self,
+                                   name,
+                                   peer_id);
+                            self.peer_mgr.correct_routing_state_to_direct(&peer_id);
+                        } else {
+                            debug!("{:?} Should have a tunnel connection to {} {:?}, but no \
+                                    tunnel node or direct connection exists.",
+                                   self,
+                                   name,
+                                   peer_id);
+                            peer_ids_to_drop.push(peer_id);
+                        }
+                    }
+                }
+            } else if !self.crust_service.is_connected(&peer_id) {
+                error!("{:?} Should have a direct connection to {} {:?}, but don't.",
+                       self,
+                       name,
+                       peer_id);
+                peer_ids_to_drop.push(peer_id);
+            }
+        }
+        let mut transition = Transition::Stay;
+        for peer_id in peer_ids_to_drop {
+            if let Transition::Terminate = self.handle_lost_peer(peer_id, outbox) {
```

Comment:
We only add the `PeerId` of a tunnel client in here if the tunnel node doesn't exist in `tunnels` and if we're not directly-connected to the other tunnel client.  In that case, there is no tunnel to close.